src/instructlab/training/utils.py

```diff
@@ -624,71 +624,114 @@ def _copy_no_lora_dict(state_dict):
-        if is_lora:
-            model_to_save.merge_adapter()
+        output_dir = final_output_dir
 
-        model_state = model_to_save.state_dict()
+    CONFIG_NAME = "config.json"
+    output_config_file = output_dir / CONFIG_NAME
+
+    get_state_dict_unpatched = accelerator.get_state_dict
+
+    def _get_state_dict_patched(model, unwrap=False):
+        return get_state_dict_unpatched(model, unwrap=unwrap)
+
+    accelerator.get_state_dict = _get_state_dict_patched
+
+    if accelerator.is_main_process:
         if is_lora:
-            model_state = _copy_no_lora_dict(model_state)
-        output_dir.mkdir(parents=True, exist_ok=True)
-        output_model_file = output_dir / WEIGHTS_NAME
-        output_config_file = output_dir / CONFIG_NAME
+            model.module.merge_adapter()
+            model_state = model.module.state_dict()
 
-        tmp_conf = copy(model_to_save.config)
-        if not tmp_conf.architectures:
-            tmp_conf.architectures = ["LlamaForCausalLM"]
+        output_dir.mkdir(parents=True, exist_ok=True)
+        if not model.module.config.architectures and convert_granite:
+            model.module.config.architectures = ["LlamaForCausalLM"]
             warnings.warn(
-                f"No architectures provided in base model config, adding architectures to ckpt: {tmp_conf.architectures}",
+                f"Adding architectures to ckpt: {model.module.config.architectures}",
             )
-
-        if args.is_granite and convert_granite:
-            with TemporaryDirectory("w") as tmpdir:
-                save_file(model_state, Path(tmpdir) / WEIGHTS_NAME)
-                tmp_conf.to_json_file(Path(tmpdir) / CONFIG_NAME)
-                tokenizer.save_pretrained(tmpdir)
-                # export doesn't like the directory to exist
-                shutil.rmtree(output_dir)
-
-                export_to_huggingface(
-                    pretrained_model_name_or_path=tmpdir,
-                    save_path=output_dir,
-                    model_type=MODEL_TYPE,
-                )
-        else:
-            torch.save(model_state, str(output_model_file))
-            tmp_conf.to_json_file(str(output_config_file))
-            tokenizer.save_pretrained(str(output_dir))
+        model.module.config.to_json_file(output_config_file)
+        tokenizer.save_pretrained(output_dir)
 
         if is_lora:
-            model_to_save.unmerge_adapter()
+            save_dict_accelerate(
+                accelerator,
+                model_state,
+                save_directory=output_dir,
+                max_shard_size="10GB",
+                safe_serialization=True,
+            )
+            model.module.unmerge_adapter()
 
-    dist.barrier()
-    log_rank_0(f"\033[93mModel saved in {output_dir}\033[0m", to_print=True)
+    if not is_lora:
+        accelerator.save_model(
+            model,
+            save_directory=output_dir,
+            max_shard_size="10GB",
```

Comment:
Let's lower the `max_shard_size` here to be consistent with the checkpoints we have published.

```suggestion
            max_shard_size="5B",
```

Reply:
*5GB, 5B would be a lot of shards lol

Reply:
agreed haha